File: lib/network/packets/server/s_pc_pack.dart
/// 同一張地圖上的玩家（含收件者自己）。
///
/// 伺服器對每張地圖只算一次名單，所以**自己也在裡面**，由前端濾掉。
///
/// 這一包是「整份名單」而不是增量的新增／移除 —— 只要地圖上的人有變動
/// 就整包重送。同圖人數是個位數，這樣可以整類消滅「畫面留下鬼影」的問題。
class SPcPack {
  const SPcPack({required this.mapId, required this.players});

  final int mapId;
  final List<RemotePlayerData> players;

  factory SPcPack.fromData(Map<String, dynamic> data) => SPcPack(
        mapId: (data['mapId'] as num?)?.toInt() ?? 0,
        players: ((data['players'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(RemotePlayerData.fromJson)
            .toList(growable: false),
      );
}

/// 名單裡的一個人。
class RemotePlayerData {
  const RemotePlayerData({
    required this.objId,
    required this.name,
    required this.x,
    required this.y,
    required this.heading,
    required this.sex,
    required this.level,
  });

  final int objId;
  final String name;
  final int x;
  final int y;
  final int heading;

  /// 0=男 1=女。外觀目前只由這個決定。
  final int sex;
  final int level;

  factory RemotePlayerData.fromJson(Map<String, dynamic> j) => RemotePlayerData(
        objId: (j['objId'] as num?)?.toInt() ?? 0,
        name: j['name'] as String? ?? '',
        x: (j['x'] as num?)?.toInt() ?? 0,
        y: (j['y'] as num?)?.toInt() ?? 0,
        heading: (j['heading'] as num?)?.toInt() ?? 2,
        sex: (j['sex'] as num?)?.toInt() ?? 0,
        level: (j['level'] as num?)?.toInt() ?? 1,
      );
}

File: src/main/java/com/xin/server/packet/server/S_PcPack.java
package com.xin.server.packet.server;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.xin.server.model.instance.PcInstance;
import com.xin.server.network.Client;
import com.xin.server.network.ClientManager;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;

/**
 * 某張地圖上的其他玩家（對應 opcode {@code S_PC_PACK}）。
 * <p>
 * <b>整包重送，不做增量。</b>只要地圖上的人變動（進圖、換圖、下線）就把
 * 該圖的完整名單重送給圖上的每個人。同圖人數是個位數，封包很小，而這樣
 * 可以整類消滅「新增與移除對不上、畫面留下鬼影」的問題 —— 與
 * {@code PartyManager} 對隊伍的處理是同一個取捨。
 * <p>
 * 名單<b>包含收件者自己</b>由前端負責濾掉：伺服器對每張地圖只算一次名單，
 * 而不是為每個人各做一份少了自己的版本。
 *
 * <pre>
 * {
 *   "op": "S_PC_PACK",
 *   "data": {
 *     "mapId": 1,
 *     "players": [
 *       { "objId": 10099, "name": "洛清塵", "x": 40, "y": 40, "heading": 2,
 *         "sex": 0, "level": 3 }
 *     ]
 *   }
 * }
 * </pre>
 */
public class S_PcPack extends ServerBasePacket {

    private S_PcPack(int mapId) {
        super(ServerOpcodes.S_PC_PACK);
        put("mapId", mapId);

        ArrayNode arr = newArray();
        for (Client c : ClientManager.getAll()) {
            if (!c.hasActiveChar()) {
                continue;
            }
            PcInstance pc = c.getActiveChar();
            if (pc.getMapId() != mapId) {
                continue;
            }
            ObjectNode node = newObject();
            node.put("objId",   pc.getId());
            node.put("name",    pc.getName());
            node.put("x",       pc.getX());
            node.put("y",       pc.getY());
            node.put("heading", pc.getHeading());
            // 外觀目前只由性別決定（前端的 appearanceKey）
            node.put("sex",     pc.getSex());
            node.put("level",   pc.getRealmLevel());
            arr.add(node);
        }
        putArray("players", arr);
    }

    public static S_PcPack of(int mapId) {
        return new S_PcPack(mapId);
    }
}
